internal/datanode/channel_manager.go

```diff
@@ -0,0 +1,505 @@
+			resp.State = datapb.ChannelWatchState_ToRelease
+			return resp
+		}
+
+		resp.State = datapb.ChannelWatchState_ReleaseFailure
+		return resp
+	default:
+		err := merr.WrapErrParameterInvalid("ToWatch or ToRelease", info.GetState().String())
+		log.Warn("fail to get progress", zap.Error(err))
+		resp.Status = merr.Status(err)
+		return resp
+	}
+}
+
+func (m *ChannelManager) Close() {
+	m.closeOnce.Do(func() {
+		m.opRunners.Range(func(channel string, runner *opRunner) bool {
+			runner.Close()
+			return true
+		})
+		m.runningFlowgraphs.close()
+		close(m.closeCh)
+		m.closeWaiter.Wait()
+	})
+}
+
+func (m *ChannelManager) Start() {
+	m.closeWaiter.Add(2)
+
+	go m.runningFlowgraphs.start(&m.closeWaiter)
+	go func() {
+		defer m.closeWaiter.Done()
+		log.Info("DataNode ChannelManager start")
+		for {
+			select {
+			case opState := <-m.communicateCh:
+				m.handleOpState(opState)
+			case <-m.closeCh:
+				log.Info("DataNode ChannelManager exit")
+				return
+			}
+		}
+	}()
+}
+
+func (m *ChannelManager) handleOpState(opState *opState) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	log := log.With(
+		zap.Int64("opID", opState.opID),
+		zap.String("channel", opState.channel),
+		zap.String("State", opState.state.String()),
+	)
+	switch opState.state {
+	case datapb.ChannelWatchState_WatchSuccess:
+		log.Info("Success to watch")
+		m.runningFlowgraphs.Add(opState.fg)
+		m.finishOp(opState.opID, opState.channel)
+
+	case datapb.ChannelWatchState_WatchFailure:
+		log.Info("Fail to watch")
+		m.finishOp(opState.opID, opState.channel)
+
+	case datapb.ChannelWatchState_ReleaseSuccess:
+		log.Info("Success to release")
+		m.finishOp(opState.opID, opState.channel)
+		m.destoryRunner(opState.channel)
+
+	case datapb.ChannelWatchState_ReleaseFailure:
+		log.Info("Fail to release, add channel to abnormal lists")
+		m.abnormals.Insert(opState.opID, opState.channel)
+		m.finishOp(opState.opID, opState.channel)
+		m.destoryRunner(opState.channel)
+	}
+}
+
+func (m *ChannelManager) getOrCreateRunner(channel string) *opRunner {
+	runner, loaded := m.opRunners.GetOrInsert(channel, NewOpRunner(channel, m.dn, m.releaseFunc, m.communicateCh))
+	if !loaded {
+		runner.Start()
+	}
+	return runner
+}
+
+func (m *ChannelManager) destoryRunner(channel string) {
+	if runner, loaded := m.opRunners.GetAndRemove(channel); loaded {
+		runner.Close()
+	}
+}
+
+func (m *ChannelManager) finishOp(opID int64, channel string) {
+	if runner, loaded := m.opRunners.Get(channel); loaded {
+		runner.FinishOp(opID)
+	}
+}
+
+type opInfo struct {
+	tickler *tickler
+}
+
+type opRunner struct {
+	channel     string
+	dn          *DataNode
+	releaseFunc releaseFunc
+
+	guard      sync.RWMutex
+	allOps     map[UniqueID]*opInfo // opID -> tickler
+	opsInQueue chan *datapb.ChannelWatchInfo
+	resultCh   chan *opState
+
+	closeWg   sync.WaitGroup
+	closeOnce sync.Once
+	closeCh   chan struct{}
+}
+
+func NewOpRunner(channel string, dn *DataNode, f releaseFunc, resultCh chan *opState) *opRunner {
```

Comment:
Move ops manager to a seperate file might be better. 
But It won't be a blocking issue